### PR TITLE
try out IERC4626

### PR DESCRIPTION
**Pull request by pandadefi**
## Description

Make Gauge IERC4626 compliant.

## Checklist

- [ ] I have run vyper and solidity linting
- [ ] I have run the tests on my machine
- [ ] I have followed commitlint guidelines
- [ ] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

File: contracts/Gauge.sol [hunk 1]
```diff
@@ -2,7 +2,9 @@
 pragma solidity 0.8.13;
 
 import "./interfaces/IExtraReward.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin-upgradeable/contracts/token/ERC20/ERC20Upgradeable.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import "@openzeppelin/contracts/utils/math/Math.sol";
```
Comment (rareweasel) [inline contracts/Gauge.sol:9]: I would also use the contracts in the `penzeppelin-contracts-upgradeable` dependency

File: contracts/Gauge.sol [hunk 2]
```diff
@@ -16,25 +18,23 @@ import "./interfaces/IVotingEscrow.sol";
     @dev this contract is used behind multiple delegate proxies.
  */
 
-contract Gauge is BaseGauge, IGauge {
+contract Gauge is BaseGauge, ERC20Upgradeable, IGauge {
```
Comment (rareweasel) [inline contracts/Gauge.sol:21]: I would create the `IERC4626` interface and use it here to make the `IERC4626 compliant` explicit for external devs/users.

File: contracts/Gauge.sol [hunk 3]
```diff
@@ -383,162 +450,246 @@ contract Gauge is BaseGauge, IGauge {
...
 
-        return true;
+    /** @notice Burns shares from owner and sends exactly assets of underlying tokens to receiver.
+     *  @dev This call updates claimable rewards
+     *  @param _assets amount to withdraw
+     *  @param _receiver account that will recieve the shares
+     *  @param _owner shares will be taken from account
+     *  @return amount of shares withdrawn
+     */
+    function withdraw(
```
Comment (rareweasel) [inline contracts/Gauge.sol:577]: I would add `_` as prefix (internal function).

**Review by rareweasel — CHANGES_REQUESTED** [review]

File: contracts/Gauge.sol [hunk 4]
```diff
@@ -16,25 +18,23 @@ import "./interfaces/IVotingEscrow.sol";
...
+contract Gauge is BaseGauge, ERC20Upgradeable, IGauge {
     using SafeERC20 for IERC20;
 
     struct Balance {
         uint256 realBalance;
         uint256 boostedBalance;
-        uint256 lastDeposit;
     }
 
     struct Appoved {
```
[the remark this answers is not in the document]
Comment (pandadefi) [reply contracts/Gauge.sol:29]: Not used anymore

File: contracts/Gauge.sol (hunk 3 above)
Comment (pandadefi) [reply contracts/Gauge.sol:577]: This should be external

File: contracts/Gauge.sol [hunk 5]
```diff
@@ -687,27 +823,28 @@ contract Gauge is BaseGauge, IGauge {
...
+            _kick(_accounts[i]);
+        }
+    }
 
+    function _kick(address _account) internal updateReward(_account) {
+        uint256 balance = balanceOf(_account);
         require(
-            balance.boostedBalance >
-                (balance.realBalance * boostingFactor) / BOOST_DENOMINATOR,
+            _boostedBalances[_account] >
```
Comment (storming0x) [inline contracts/Gauge.sol:842]: Can this logic be a view function with address argument that return true or false? if i want to check accounts that potentially are candidates for kicking?

File: contracts/Gauge.sol [hunk 6]
```diff
@@ -150,31 +149,61 @@ contract Gauge is BaseGauge, IGauge {
 
     /** @return total of the staked vault token
      */
-    function totalSupply() public view returns (uint256) {
```
Comment (storming0x) [inline contracts/Gauge.sol:153]: erc4626 is also an  erc20 so this method still needs to be available to be erc20 compliant, not sure if the erc4626 interface extends from correct erc20 interface.

File: contracts/Gauge.sol [hunk 7]
```diff
@@ -150,31 +149,61 @@ contract Gauge is BaseGauge, IGauge {
...
-        return _totalSupply;
+    function totalAssets() public view returns (uint256) {
+        return totalSupply();
     }
 
-    /** @param _account to look balance for
-     *  @return amount of staked token for an account
-     */
-    function balanceOf(address _account)
```
Comment (storming0x) [inline contracts/Gauge.sol:160]: same as above, maybe it was moved elsewhere? not seeing it.

File: contracts/Gauge.sol [hunk 8]
```diff
@@ -383,162 +435,246 @@ contract Gauge is BaseGauge, IGauge {
...
+    */
+    function maxWithdraw(address _owner) external view returns (uint256) {
+        return balanceOf(_owner);
+    }
+
+    function previewWithdraw(uint256 _assets) external view returns (uint256) {
+        return _assets;
     }
 
     /** @notice allow an address to deposit on your behalf
```
Comment (storming0x) [inline contracts/Gauge.sol:520]: no longer valid this part of the comment, can update to canClaim and  canLock

File: contracts/Gauge.sol [hunk 9]
```diff
@@ -261,15 +290,38 @@ contract Gauge is BaseGauge, IGauge {
...
+        if (_from != address(0)) {
+            _updateReward(_from);
+            //also deposit to linked rewards
+            uint256 length = extraRewards.length;
+            for (uint256 i = 0; i < length; ++i) {
+                IExtraReward(extraRewards[i]).rewardCheckpoint(_from);
+            }
+        }
+        if (_to != address(0)) {
+            _updateReward(_to);
```
Comment (storming0x) [inline contracts/Gauge.sol:307]: what about updating balance.realbalance and other struct related data in this point, it seems transfers should put it out of sync with realBalance.

Say a user deposits, later gets a transfer from a second user, balance.realbalance, balance.lastDeposit and balance.boostedBalance will be out of sync for the user, i guess not in his favor.

lastDeposit seems to make sense to update based on a transfer, basically a transfer should be equivalent to a depositFor in accounting housekeeping.

File: contracts/Gauge.sol [hunk 10]
```diff
@@ -383,162 +435,246 @@ contract Gauge is BaseGauge, IGauge {
...
+        uint256 _assets,
+        address _receiver,
+        address _owner,
         bool _claim,
         bool _lock
-    ) public updateReward(msg.sender) returns (bool) {
-        require(_amount != 0, "RewardPool : Cannot withdraw 0");
-        Balance storage balance = _balances[msg.sender];
-        require(
-            balance.lastDeposit < block.number,
```
Comment (storming0x) [inline contracts/Gauge.sol:487]: Cant we add this balance.lastDeposit check state on transfers too? that way we don't need to remove it here since its a valid guard, as pointed out from yacademy

Comment (storming0x) [reply contracts/Gauge.sol:487]: Not sure the kick function is a complete mitigation to the flashloan boost problem, its similar to the harvest sandwiches in vault and before we had lockedProfitDegradation it was a complete whack a mole game. 

Is the plan to mitigate when we send rewards to kick in the same trx? i can see us getting sandwiched when we drop rewards or after kicking someone, they can flashloan again and boost after in different account after transfering.

Basically im not sure it can be solve with lastDeposit check, may need to have degradation on deposits in the same way harvest profit also has locking and degradation.

File: contracts/Gauge.sol (hunk 9 above)
Comment (rareweasel) [reply contracts/Gauge.sol:307]: I agree,

If I can transfer my gTokens, when I transfer it to another user, we should do something similar to what we do in the `_depositFor` and `_withdraw` functions:

- [_depositFor](https://github.com/yearn/veYFI/blob/IERC4626/contracts/Gauge.sol#L493)
- [_withdraw](https://github.com/yearn/veYFI/blob/IERC4626/contracts/Gauge.sol#L632-L634)

Essentially, we update the `_boostedBalances` for the given account, but in this case I think we should do it for both addresses (`_to` and `_from`):

```
        uint256 boostedBalance = _boostedBalanceOf(_receiver);
        _boostedBalances[_receiver] = boostedBalance;

        emit BoostedBalanceUpdated(_receiver, boostedBalance);
```

A side note: I think we should add that logic in the `_afterTokenTransfer` function, because we are using `_beforeTokenTransfer` and the account balances will not be updated at that point in time to update the `boostedBalances`. Please, CMIIW.

Reference:  The [transfer](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L226-L246) function has two hooks:

- [_beforeTokenTransfer(from, to, amount)](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L234)
- [_afterTokenTransfer(from, to, amount)](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L245)

File: contracts/Gauge.sol [hunk 11]
```diff
@@ -52,15 +52,12 @@ contract Gauge is BaseGauge, IGauge {
...
     mapping(address => mapping(address => Appoved)) public approvedTo;
 
     //// @notice list of extraRewards pool.
     address[] public extraRewards;
 
-    event Staked(address indexed user, uint256 amount);
-    event Withdrawn(address indexed user, uint256 amount);
     event AddedExtraReward(address indexed reward);
     event DeletedExtraRewards(address[] rewards);
     event RemovedExtraReward(address indexed reward);
```
Comment (rareweasel) [inline contracts/Gauge.sol:63]: Maybe I am missing something but why do we have two events for similar actions?

Please, cmiiw, Is it possible to use the `DeletedExtraRewards` event and pass an array with only one item where we are using the `RemovedExtraReward` event?

File: contracts/Gauge.sol (hunk 9 above)
Comment (rareweasel) [reply contracts/Gauge.sol:307]: Also, imo we should add unit tests to cover all the potential scenarios, and test it on ropsten too.

**Review by rareweasel — CHANGES_REQUESTED** [review]

File: contracts/Gauge.sol (hunk 2 above)
Comment (pandadefi) [reply contracts/Gauge.sol:21]: IGauge is inheriting from IERC4626, isn't it enough?
```
pragma solidity 0.8.14;
import "./IBaseGauge.sol";
import "./IERC4626.sol";

interface IGauge is IBaseGauge, IERC4626 {
```

File: contracts/Gauge.sol (hunk 10 above)
Comment (pandadefi) [reply contracts/Gauge.sol:487]: It will be an issue for contracts that are depositing and transferring funds to an EOA (zap)

File: contracts/Gauge.sol (hunk 11 above)
Comment (pandadefi) [reply contracts/Gauge.sol:63]: That can work

File: contracts/Gauge.sol (hunk 9 above)
Comment (pandadefi) [reply contracts/Gauge.sol:307]: https://github.com/yearn/veYFI/pull/155/commits/bd5a0069001f00fb8b9df5763a05bd8057bfd8d0

File: contracts/Gauge.sol (hunk 11 above)
Comment (pandadefi) [reply contracts/Gauge.sol:63]: That will need some memory manipulation

File: contracts/Gauge.sol (hunk 10 above)
Comment (storming0x) [reply contracts/Gauge.sol:487]: > It will be an issue for contracts that are depositing and transferring funds to an EOA (zap)

you mean to track lastDeposit, yeah i don't see it as a fully complete solution because of the DDOS problem, if we don't allow users to withdraw based on lastDeposit someone can flood you with dust your account to front run lastDeposit and keep you from withdrawing, still not sure about another alternative.

File: contracts/Gauge.sol (hunk 2 above)
Comment (rareweasel) [reply contracts/Gauge.sol:21]: > IGauge is inheriting from IERC4626, isn't it enough?
> 
> ```
> pragma solidity 0.8.14;
> import "./IBaseGauge.sol";
> import "./IERC4626.sol";
> 
> interface IGauge is IBaseGauge, IERC4626 {
> ```

Good point, sorry, I missed that one.

File: contracts/Gauge.sol (hunk 11 above)
Comment (rareweasel) [reply contracts/Gauge.sol:63]: > Maybe I am missing something but why do we have two events for similar actions?
> 
> Please, cmiiw, Is it possible to use the `DeletedExtraRewards` event and pass an array with only one item where we are using the `RemovedExtraReward` event?

I think that it is more readable, and we can remove the `RemovedExtraReward` event.

File: contracts/Gauge.sol (hunk 5 above)
Comment (pandadefi) [reply contracts/Gauge.sol:851]: They all are since quickly after a lock/deposit your veYFI token decreases.

File: contracts/BaseGauge.sol [hunk 12]
```diff
@@ -3,11 +3,10 @@ pragma solidity 0.8.13;
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/utils/math/Math.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
```
[the remark this answers is not in the document]
Comment (storming0x) [reply contracts/BaseGauge.sol:5]: if the code in contracts works the same as upgradable not sure why we need to import it, it brings confusion for folk thinking these contracts are upgradable, when they are not really.

File: contracts/Gauge.sol (hunk 5 above)
Comment (storming0x) [reply contracts/Gauge.sol:851]: Dont understand your answer here regarding my question, whether this can be a view helper method we can check off chain or not.

File: contracts/Gauge.sol [hunk 13]
```diff
@@ -383,162 +448,242 @@ contract Gauge is BaseGauge, IGauge {
...
+    }
+
+    /** @notice allow an address to lock and claim on your behalf
+     * claim ermai
      *  @param _addr address to change approval for
-     *  @param _canDeposit can deposit
      *  @param _canClaim can deposit
      *  @return true
      */
     function setApprovals(
```
Comment (storming0x) [inline contracts/Gauge.sol:536]: i think it's better UX to have granular methods, instead of a single setter for boolean arguments where one argument can easily be sent in bad order and confusing for caller.

`setApprovals(0x01, true, false)` can have different consequences than `setApprovals(0x01, false, true)`

potentially error prone, i prefer explicit methods.

`enableClaimingFrom(address)`. `disableClaimingFrom(address)`

`enableLockingFrom(address)`. `disableLockingFrom(address)`

less arguments and makes the side effect explicit avoiding potential oops factor

**Review by storming0x — CHANGES_REQUESTED** [review]
Left a small comment around improving approval UX methods and adding testing.

File: contracts/Gauge.sol (hunk 9 above)
Comment (storming0x) [reply contracts/Gauge.sol:307]: LAtest commit looks good, but needs to add the unit tests around transfer use case and reward checks, transfering several times in a loop back and forth between two accounts and see if rewards are consistent.

**Comment by storming0x** [comment]
Did a review on latest changes focusing more on attack vectors around principal vault token, and malicious user can get a hold of more gauge shares than they have, could not identified any specific attack to withdraw more capital than deposited on my last review.

Regarding rewards given that the design change introduces a transfer mechanism that needs accounting logic too, im less certain theres no way to game rewards in favor of a user, we suggested adding some test cases around possible transfer attacks to validate this.

Also not entirely certain the kick function can be enough mitigation for stopping abuse of boosted rewards from flashloan attacks, it may be needed to control deposit limits around intended gauge assets at the vault level to mitigate this particular "flash boost" attack for gaining more rewards.

**Review by rareweasel — CHANGES_REQUESTED** [review]
